In [1]:
import csv
f = open('guns.csv')
csvreader = csv.reader(f)
data = list(csvreader)
# Display the 5 first rows of the data list
data[0:5]


[['',
  'year',
  'month',
  'intent',
  'police',
  'sex',
  'age',
  'race',
  'hispanic',
  'place',
  'education'],
 ['1',
  '2012',
  '01',
  'Suicide',
  '0',
  'M',
  '34',
  'Asian/Pacific Islander',
  '100',
  'Home',
  '4'],
 ['2', '2012', '01', 'Suicide', '0', 'F', '21', 'White', '100', 'Street', '3'],
 ['3',
  '2012',
  '01',
  'Suicide',
  '0',
  'M',
  '60',
  'White',
  '100',
  'Other specified',
  '4'],
 ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', '4']]

In [2]:
# Remove headers from the data list(list of lists)
headers = data[0]
data.remove(headers)
headers
# Display the first 5 rows of data to verify that the headers were removed
data[0:5]

[['1',
  '2012',
  '01',
  'Suicide',
  '0',
  'M',
  '34',
  'Asian/Pacific Islander',
  '100',
  'Home',
  '4'],
 ['2', '2012', '01', 'Suicide', '0', 'F', '21', 'White', '100', 'Street', '3'],
 ['3',
  '2012',
  '01',
  'Suicide',
  '0',
  'M',
  '60',
  'White',
  '100',
  'Other specified',
  '4'],
 ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', '4'],
 ['5',
  '2012',
  '02',
  'Suicide',
  '0',
  'M',
  '31',
  'White',
  '100',
  'Other specified',
  '2']]

In [13]:
# Counting gun deaths by year. First, use list comprehension to extract the year column from data
# Secondly, create an empty dictionary and loop through each element in years
years = [item[1] for item in data]
year_counts = {}
for year in years:
    if year not in year_counts:
        year_counts[year] = 0
    else:
        year_counts[year] += 1
year_counts

{'2012': 33562, '2013': 33635, '2014': 33598}

In [4]:
# In this step, I will analyze the guns death by months and years. For this purpose,
# is necessary to create a datetime.datetime object for each row. I use list comprehension to create a datetime.datetime object with the year and the month(day fixed)
import datetime
dates = [datetime.datetime(year=int(item[1]), month = int(item[2]), day=1) for item in data]
# Display the first 5 rows in dates 
dates[0:5]

# Create a dictionary with the datetime.datetime(year,month,fixed day) as a key and the number of deaths in each month as the value
date_counts = {}
for date in dates:
    if date not in date_counts:
        date_counts[date] = 0
    else:
        date_counts[date] += 1
date_counts

{datetime.datetime(2012, 1, 1, 0, 0): 2757,
 datetime.datetime(2012, 2, 1, 0, 0): 2356,
 datetime.datetime(2012, 3, 1, 0, 0): 2742,
 datetime.datetime(2012, 4, 1, 0, 0): 2794,
 datetime.datetime(2012, 5, 1, 0, 0): 2998,
 datetime.datetime(2012, 6, 1, 0, 0): 2825,
 datetime.datetime(2012, 7, 1, 0, 0): 3025,
 datetime.datetime(2012, 8, 1, 0, 0): 2953,
 datetime.datetime(2012, 9, 1, 0, 0): 2851,
 datetime.datetime(2012, 10, 1, 0, 0): 2732,
 datetime.datetime(2012, 11, 1, 0, 0): 2728,
 datetime.datetime(2012, 12, 1, 0, 0): 2790,
 datetime.datetime(2013, 1, 1, 0, 0): 2863,
 datetime.datetime(2013, 2, 1, 0, 0): 2374,
 datetime.datetime(2013, 3, 1, 0, 0): 2861,
 datetime.datetime(2013, 4, 1, 0, 0): 2797,
 datetime.datetime(2013, 5, 1, 0, 0): 2805,
 datetime.datetime(2013, 6, 1, 0, 0): 2919,
 datetime.datetime(2013, 7, 1, 0, 0): 3078,
 datetime.datetime(2013, 8, 1, 0, 0): 2858,
 datetime.datetime(2013, 9, 1, 0, 0): 2741,
 datetime.datetime(2013, 10, 1, 0, 0): 2807,
 datetime.datetime(2013, 11,

In [15]:
# In this step, I will explore gun deaths by race and sex
# First, I create a dictionary to count the deaths gun by gender
gender = [item[5] for item in data]
sex_counts = {}
for item in gender:
    if item not in sex_counts:
        sex_counts[item] = 0
    else:
        sex_counts[item] += 1

# Secondly, create a dictionary to count the deaths by race
races = [item[7] for item in data]
race_counts = {}
for item in races:
    if item not in race_counts:
        race_counts[item] = 0
    else:
        race_counts[item] += 1
print(sex_counts)
print(race_counts)

{'F': 14448, 'M': 86348}
{'Asian/Pacific Islander': 1325, 'Native American/Native Alaskan': 916, 'Black': 23295, 'White': 66236, 'Hispanic': 9021}


The result yields interesting conclusion. In the first hand, there are almost 6 times more deaths in males than females. 

Secondly, among the race of the deaths, the white people are the most killed, with a number of deaths of 66237 between 2012 to 2014. This number is almost three times higher than the black deaths that are 23296. 

I consider that it will be intereting to examinate the age of the people killed from the males population. The male population who deaths in these years, represents more than 85 % of the sample. 

Also is important to analyze the race deaths, in releative terms. This imply to take into account the proportion of deaths of each reace in terms of the total population of each race in the US.



In [6]:
# In order to know the rate of gun deaths per 100000 people of each race, it's necessary to read the Census Database to know what percentage of the US population falls into each racial category.
# I will use the census population dataset, that contain information about the population of each race in the United States.
f = open('census.csv')
csvreader = csv.reader(f)
census = list(csvreader)
census

[['Id',
  'Year',
  'Id',
  'Sex',
  'Id',
  'Hispanic Origin',
  'Id',
  'Id2',
  'Geography',
  'Total',
  'Race Alone - White',
  'Race Alone - Hispanic',
  'Race Alone - Black or African American',
  'Race Alone - American Indian and Alaska Native',
  'Race Alone - Asian',
  'Race Alone - Native Hawaiian and Other Pacific Islander',
  'Two or More Races'],
 ['cen42010',
  'April 1, 2010 Census',
  'totsex',
  'Both Sexes',
  'tothisp',
  'Total',
  '0100000US',
  '',
  'United States',
  '308745538',
  '197318956',
  '44618105',
  '40250635',
  '3739506',
  '15159516',
  '674625',
  '6984195']]

In [14]:
# Combine the two datasets to calculate the rate of deaths among each racial category. The issue in this point is that the racial categories are named
# differently in census respect to the data list. It will be necessary to construct a dictionary to map between them and perform the division, to get the rate of gun deaths for each racial category.
mapping = {'Asian/Pacific Islander': 15834141, 'Black': 40250635, 'Hispanic': 44618105, 'Native American/Native Alaskan': 3739506, 'White': 197318956}

# Divide the number of deaths for each race, to the population of each race in the US and multiply by 100000.
race_per_hundredk = {}
for key in race_counts:
    race_per_hundredk[key] = race_counts[key]/mapping[key] * 100000
race_per_hundredk

{'Asian/Pacific Islander': 8.36799419684339,
 'Black': 57.87486334066531,
 'Hispanic': 20.2182499682584,
 'Native American/Native Alaskan': 24.49521407373059,
 'White': 33.567986240511026}

These conclusion, contrast to the figures that only take the numbers of gun deaths in each race without taking into account  the population of each race. In relative terms, Black people has the most rate of gun deaths with 58 out of every 100000 people in the Black census category in the US between 2012 and 2014.

In [8]:
# In this step, I will perform the same analysis of the last cell, but only count rows where the intent was Homicide 
# First extract the intent and races column using list comprehension
intents = [item[3] for item in data]

# Extract the races column using list comprehension
races = [item[7] for item in data]

# Using enumerate function, to loop over the races column and count the number of Homicide for each race.
homicide_race_counts = {}
for i, race in enumerate(races):
    if race not in homicide_race_counts:
        homicide_race_counts[race] = 0
    if intents[i] == 'Homicide':
        homicide_race_counts[race] += 1    
              
print(homicide_race_counts)

# Create a dictionary to find the rate of Homicide by the population of each race 
homicide_race_per_hundredk = {}
for key in homicide_race_counts:
    homicide_race_per_hundredk[key] = homicide_race_counts[key] /  mapping[key] * 100000
print(homicide_race_per_hundredk)    

                

{'Asian/Pacific Islander': 559, 'Native American/Native Alaskan': 326, 'Black': 19510, 'White': 9147, 'Hispanic': 5634}
{'Asian/Pacific Islander': 3.530346230970155, 'Native American/Native Alaskan': 8.717729026240365, 'Black': 48.471284987180944, 'White': 4.6356417981453335, 'Hispanic': 12.627161104219914}


The results shows that black people has the most rate of Homicide in terms of the population. This figure depict that 48 out of every 100000  people in the Black census race category in the US were killed by Homicide between 2012 to 2014.

On the other hand, Asian/Pacific Islander race is the less killed by Homicide. The White race also have a less rate of people killed by Homicide with 5 out of ecery 100000 people in the White census race category in the US were killed by guns between 2012 and 2014. 

It's seems that white people die by other causes instead of homicide, such as accidental causes or suicide. I will explore this in next steps.

In [19]:
# Examine if there are any link between month and homicide rate
import operator 
months = [item[2] for item in data]
intents = [item[3] for item in data]

homicide_by_month = {}

for i,month in enumerate(months):
    if month not in homicide_by_month:
        homicide_by_month[month] = 0
    if intents[i] == 'Homicide':
        homicide_by_month[month] += 1
print(homicide_by_month)

# I construct a dictionary(map_division) to map with the homicide_by_month dictionary and generate a homicide index in each month. The index has base 100 and start in month 01.
# The dictionary homicide_by_month_rate has a key of each month and the value is an Index on base 100, that reflect the increment of homicide respect the month base which is 01.
map_division = {'01': 2829, '09': 2829, '12': 2829, '04': 2829, '02': 2829 ,'11': 2829, '10': 2829, '08': 2829 , '05': 2829 ,'03': 2829 ,'06': 2829, '07': 2829}

index_homicide_by_month = {}
for k in homicide_by_month:
    if k not in index_homicide_by_month:
        index_homicide_by_month[k] = round((homicide_by_month[k]/ map_division[k]) * 100,3)
    if homicide_by_month.keys == homicide_by_month_rate.keys:
        index_homicide_by_month[k] = round((homicide_by_month[k]/ map_division[k]) * 100,3)
print(index_homicide_by_month)

{'04': 2845, '06': 3130, '05': 2976, '03': 2780, '08': 3125, '09': 2966, '01': 2829, '12': 3191, '10': 2968, '07': 3269, '02': 2178, '11': 2919}
{'04': 100.566, '08': 110.463, '05': 105.196, '02': 76.988, '06': 110.64, '09': 104.843, '01': 100.0, '12': 112.796, '10': 104.913, '07': 115.553, '03': 98.268, '11': 103.181}


The results show that in the first half of the year, the rate of homicide measured by the index_homicide is less respect to the second part of the year. The month in which the rate of homicide has the highest index is July followed by December. 

In [10]:
# In this step, I analyze the relationship between the homicide rate by gender
gender = [item[5] for item in data]
gender_homicide = {}
for i, sex in enumerate(gender):
    if sex not in gender_homicide:
        gender_homicide[sex] = 0
    if intents[i] == 'Homicide':
        gender_homicide[sex] += 1
print(gender_homicide)

# Create a dictionary to obtain the population by gender. The information provided by the 2010 Census, yield that the 49% of the population are men and the 51% are woman.
# With this information, I can divide the homicide be gender to the population of each gender and get a rate of homicide by gender
population= {'M': 151960127 , 'F': 157460221}
gender_homicide_rate = {}
for key in gender_homicide:
    gender_homicide_rate[key] = (gender_homicide[key]/population[key]) * 100000
gender_homicide_rate

{'F': 5373, 'M': 29803}


{'F': 3.412290396823462, 'M': 19.61238160849918}

The result show that 20 out of 100000 men have death by homicide between 2012 to 2014. This figure is significantly smaller for woman, with only 3 woman out of 100000 was death by homicide between 2012 to 2014. 

In [11]:
# In this step I try fo find out if there are correlations between gun death rates and location or education
# At first time, I construct a diccionary to show up if the education of the victimis has any correlation with the numbers of death
education = [item[10] for item in data]

gun_death_by_education = {}
for item in education:
    if item not in gun_death_by_education:
        gun_death_by_education[item] = 0
    else:
        gun_death_by_education[item] += 1
gun_death_by_education

{'1': 21822, '2': 42926, '3': 21679, '4': 12945, '5': 1368, 'NA': 52}

The result show that the highest death fall in the category 2 of education which is 'Graduated from High School or equivalent'. 
It seems to be that people that go the University or finish it, have less number of death respect to people that went to High School but didn't finish or are graduated from High School.

In [12]:
# Next, I observe if there are correlations between location and death rates
# The methodology will be the same that before. First, extract the location item from data, and then loop over the location item to make a dictionary where the keys
# are the location and the values the number of death in each location.
location = [item[9] for item in data]
death_by_location = {}
for item in location:
    if item not in death_by_location:
        death_by_location[item] = 0
    else:
        death_by_location[item] += 1
death_by_location

{'Farm': 469,
 'Home': 60485,
 'Industrial/construction': 247,
 'NA': 1383,
 'Other specified': 13750,
 'Other unspecified': 8866,
 'Residential institution': 202,
 'School/instiution': 670,
 'Sports': 127,
 'Street': 11150,
 'Trade/service area': 3438}

The conclusion of the results are that, the highest number of death take place in Home with 60485. This number is really high respect the follwing place, that is Street with 11150 deaths. 

The location with the lowest number of deaths between 2012 to 2014 is Sports with 127 deaths . 

In [24]:
# Finally I will analyze the Suicide cause of death and find out the relationship between Suicide causes of death and race.

intents = [item[3] for item in data]

# Extract the races column using list comprehension
races = [item[7] for item in data]

# Using enumerate function, to loop over the races column and count the number of Suicide for each race.
suicide_race_counts = {}
for i, race in enumerate(races):
    if race not in suicide_race_counts:
        suicide_race_counts[race] = 0
    if intents[i] == 'Suicide':
        suicide_race_counts[race] += 1    
              
print(suicide_race_counts)

# Create a dictionary to find the rate of Suicide by the population of each race 
suicide_race_per_hundredk = {}
for key in suicide_race_counts:
    suicide_race_per_hundredk[key] = suicide_race_counts[key] /  mapping[key] * 100000
print(suicide_race_per_hundredk) 

{'Asian/Pacific Islander': 745, 'Native American/Native Alaskan': 555, 'Black': 3332, 'White': 55372, 'Hispanic': 3171}
{'Asian/Pacific Islander': 4.705023152187416, 'Native American/Native Alaskan': 14.841532544673013, 'Black': 8.278130270491385, 'White': 28.06217969245692, 'Hispanic': 7.106980451097149}


The results show that white people is the category with the highest rate of suicide as a cuase of die. The second race with the most rate of suicide death is Native American/Native Alaskan race. 

Black people has a low rate with only 8 out of every 100000 black people were death by suicide between 2012 and 2014. Such as the results of homicide rate show, for black people the main cause of die is homicide. 